### DISCUSSION TITLE
time_t

### DISCUSSION THREAD

**Discussion by XosRod** [jpb10/SolarCalculator#26] [Q&A]
hello.
I have a little problem: 
all the examples where "time_t" appears give the following error:

/home/xose/PROXECTOS/SeguidorSolar/SolarCalculator-main(!!)/examples/SolarCalculatorTimeLib/SolarCalculatorTimeLib.ino:68:1: error: 'time_t' does not name a type; did you mean 'size_t'?
 time_t toUtc(time_t local)
 ^~~~~~
 size_t
/home/xose/PROXECTOS/SeguidorSolar/SolarCalculator-main(!!)/examples/SolarCalculatorTimeLib/SolarCalculatorTimeLib.ino:79:1: error: 'time_t' does not name a type; did you mean 'size_t'?
 time_t compileTime()
 ^~~~~~
 size_t
/home/xose/PROXECTOS/SeguidorSolar/SolarCalculator-main(!!)/examples/SolarCalculatorTimeLib/SolarCalculatorTimeLib.ino: In function 'void setup()':
/home/xose/PROXECTOS/SeguidorSolar/SolarCalculator-main(!!)/examples/SolarCalculatorTimeLib/SolarCalculatorTimeLib.ino:27:17: error: 'compileTime' was not declared in this scope
   setTime(toUtc(compileTime()));
                 ^~~~~~~~~~~
/home/xose/PROXECTOS/SeguidorSolar/SolarCalculator-main(!!)/examples/SolarCalculatorTimeLib/SolarCalculatorTimeLib.ino:27:11: error: 'toUtc' was not declared in this scope
   setTime(toUtc(compileTime()));
           ^~~~~
/home/xose/PROXECTOS/SeguidorSolar/SolarCalculator-main(!!)/examples/SolarCalculatorTimeLib/SolarCalculatorTimeLib.ino:27:3: error: 'setTime' was not declared in this scope
   setTime(toUtc(compileTime()));
   ^~~~~~~
/home/xose/PROXECTOS/SeguidorSolar/SolarCalculator-main(!!)/examples/SolarCalculatorTimeLib/SolarCalculatorTimeLib.ino:33:3: error: 'time_t' was not declared in this scope
   time_t utc = now();
   ^~~~~~
/home/xose/PROXECTOS/SeguidorSolar/SolarCalculator-main(!!)/examples/SolarCalculatorTimeLib/SolarCalculatorTimeLib.ino:33:3: note: suggested alternative: 'size_t'
   time_t utc = now();
   ^~~~~~
   size_t
/home/xose/PROXECTOS/SeguidorSolar/SolarCalculator-main(!!)/examples/SolarCalculatorTimeLib/SolarCalculatorTimeLib.ino:35:22: error: 'utc' was not declared in this scope
   calcEquationOfTime(utc, eq);
                      ^~~
/home/xose/PROXECTOS/SeguidorSolar/SolarCalculator-main(!!)/examples/SolarCalculatorTimeLib/SolarCalculatorTimeLib.ino:35:22: note: suggested alternative: 'putc'
   calcEquationOfTime(utc, eq);
                      ^~~
                      putc
/home/xose/PROXECTOS/SeguidorSolar/SolarCalculator-main(!!)/examples/SolarCalculatorTimeLib/SolarCalculatorTimeLib.ino: At global scope:
/home/xose/PROXECTOS/SeguidorSolar/SolarCalculator-main(!!)/examples/SolarCalculatorTimeLib/SolarCalculatorTimeLib.ino:68:1: error: 'time_t' does not name a type; did you mean 'size_t'?
 time_t toUtc(time_t local)
 ^~~~~~
 size_t
/home/xose/PROXECTOS/SeguidorSolar/SolarCalculator-main(!!)/examples/SolarCalculatorTimeLib/SolarCalculatorTimeLib.ino:79:1: error: 'time_t' does not name a type; did you mean 'size_t'?
 time_t compileTime()
 ^~~~~~
 size_t

exit status 1

Compilation error: 'time_t' does not name a type; did you mean 'size_t'?

how could i solve it?

**Comment by jpb10** [comment]
For this particular example, make sure that the [Time](https://github.com/PaulStoffregen/Time) library is installed.

**Reply by XosRod** [reply]
hello, thank you for replying.
I have installed the TimeLib-2.1.0 library.
is it valid?

**Comment by XosRod** [comment]
sorry.
I have switched to the time library and the sketch already compiles.
thanks

**Comment by XosRod** [comment]
I have tested the library as it comes and it works fine, but if I change the date and time to manual and enter the data myself the azimuth and altitude results are very very wrong with respect to the NOAA ones (about 30º).
I don't understand

**Reply by jpb10** [reply]
Please post your code, input values and program output. 

Some common sources of error: 
- Input time **not in UTC**
- Wrong sign for longitude
- Not adding the UTC offset to the sunrise and sunset times

**Comment by XosRod** [comment]
CODIGO:#include <SolarCalculator.h>
#include <TimeLib.h>

// Location
double latitude = 42.1654743206;
double longitude = -8.5030256541;
int utc_offset = +2;

void setup()
{
  Serial.begin(9600);

  double transit, sunrise, sunset;  // Event times, in hours (UTC)
  double eq;                        // Equation of time, in minutes
  double ra, dec, r;                // Equatorial coordinates, in degrees and AUs
  double az, el;                    // Horizontal coordinates, in degrees

  // Set system time to compile time
  //setTime(toUtc(compileTime()));

  // Set time manually (hr, min, sec, day, mo, yr)
  setTime(16, 00, 00, 15, 10, 2023);

  // Get current time
  time_t utc = now();

SALIDA:
Sunrise: 08:46
Transit: 14:20
Sunset: 19:53
Eq of time: 14.21 min
RA: 13.359h Dec: -8.58º R: 0.997200 AU
Az: 238.80º El: 18.69º

Salida NOAA
Az: 217.65º El: 31.58º

**Comment by XosRod** [comment]
if setTime(14,00,00,15,10,2023)
the output is:
Sunrise: 08:46
Transit: 14:20
Sunset; 19:53
Eq of time: 14.19 min
RA: 13.353h Dec: -8.55º R; 0.997224 AU
Az: 210.47º El: 34.37º

**Comment by XosRod** [comment]
the latter is quite consistent with the NOAA output.

**Reply by jpb10** [reply]
It should agree with NOAA within rounding error, or 0.01 degree.

**Comment by XosRod** [comment]
it is evident that the error was in not setting the setTime in UTC

**Comment by XosRod** [comment]
thank you for your work
